src/libcore/iter.rs

```diff
@@ -369,6 +369,25 @@ pub trait Iterator {
     /// `None` here means that either there is no known upper bound, or the
     /// upper bound is larger than `usize`.
     ///
+    /// # Safety
+    ///
+    /// Implementations of this trait are not guaranteed to yield the
+    /// declared number of elements. The iterator may yield less than
+    /// lower bound or more than upper bound of elements.
```

Comment:
But if this happens, it's an implementation bug, so can this be reworded or toned down? Don't want to send a signal that being wrong in the implementation is ok.

Note that the default implementation returning `(0, None)` is correct for all iterators.